[RUNTIME][GOLANG] TVM runtime for golang v0.1

srkreddy1238 opened the pull request:
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

tqchen commented:
It is a good start, I would recommend to not use swig but directly use the C FFI in Go, and do a bit wrapping to support Module and TVMValue class like in jvm,

c.f. https://github.com/jdeng/gomxnet

srkreddy1238 commented:
@tqchen 

Ref. gomxnet above 
I see it implements a new API layer for go instead of mxnet original. Do we define a new API layer for go?

Reason I choose swig was to keep the interface maintenance automatic for changes to c_runtime_api.
And both ways here use same unsafe and C packages below.

pls advice.

srkreddy1238 commented:
@tqchen 

I checked the tvm4j and tried to wrap few types inside TVMvalue. 

I see other challenges in wrapping the go slices to C arrays for TVMValue array itself and other primitive types. golang use slices very common instead of arrays which complicates the interface. 

Hand coding slice to a array for all these types via "C" and "unsafe" need more effort and results almost same like the swig generated interface.

Please advice.

tqchen commented:
I think if we can build a  NDArray.copyfrom(go_slice) function, it will work for most case. The data conversion happens inside the function itself, so it is invisible to the users

srkreddy1238 commented:
@tqchen   please have a look.
If the code structure and API naming are good I will finish the few pending API and conclude.

@dmlc/tvm-reviewer welcome to review and comment.

tqchen commented:
@srkreddy1238 Can you do additional wrappings to provide the following class object?

- TVMFunction
- TVMModule
- NDArray
- TVMValue(for handling return value conversion)

srkreddy1238 commented:
Sure, I will add above class objects.

tqchen commented:
Also please add appropriate documentation for user-facing objects and methods, as per https://blog.golang.org/godoc-documenting-go-code

srkreddy1238 commented:
@tqchen 

- Added wrappers for TVMModule, TVMFunction
- Dropped DLTensor and given TVMArray instead.
- TVMValue getter/setter for all these types.
- Documentation as per the standard.
- golint addition for Make script.

srkreddy1238 commented:
@tqchen  ready for review.

Recent changes 
- Simplify the getter and setter on TVMValue user interface.
- GetFunction and FuncCall are simplified TVMFuncExec which take funcname and variable args.

tqchen commented:
I made some simple comments, but the wrapping is still a bit raw(it reads more like C rather than Go), here is what we want ideally(note I am not expert in Go, so syntax might be wrong)

```go
mod = gotvm.LoadFromFile("/path/to/mymodule.so")
f1 = mod.GetFunction("myfunction")
# need to debate how to support function invocation, either in tvm style or java style
f1.invoke(args) 


myarr = gotvm.empty(shape, ctx);

```

tqchen commented:
Actually since, this involves API design discussion, can you please open an RFC issue so that more people can participate? Maybe there are good ideas from the community, I am not expert in Go, so I can only generalize from my experience in other languages

srkreddy1238 commented on go/README.md:
```diff
@@ -0,0 +1,51 @@
+# tvmgo - Golang Frontend for TVM Runtime
```
Done

srkreddy1238 commented on go/runtime/gotvm.h:
```diff
@@ -0,0 +1,72 @@
+#include <tvm/runtime/c_runtime_api.h>
+#include <dlpack/dlpack.h>
+
+// Some type devinitions for golang "C"
```
gotvm.h ideally should be a command followed by ```import "C"``` 
cgo picks up picks it it from commented code.

No problem, extern works here well.

Done

srkreddy1238 commented:
#1559 for a discussion with team.

srkreddy1238 commented:
@tqchen 

```golang
f1 = mod.GetFunction("myfunction")
f1.invoke(args) 
f1.free()
```
The new API ```TVMFunctionExec``` ideally does all these. I could keep as above based on the duscussion result.

tqchen commented:
Some structural comments:

- tvm/go -> tvm/golang
- tvm/golang/runtime/src -> tvm/golang/src  (since this is the only module we advocate)
- put testcases under tvm/tests/golang

tqchen commented:
I am not expert in golang, so we will also need to ask another person to do a code review

srkreddy1238 commented:
Structural changes done.

tests can't be moved completely away as these are package internal test cases should be along with source to build and run.

Ref. https://golang.org/pkg/testing/

srkreddy1238 commented:
@dmlc/tvm-team Anybody in the team familiar with golang for a review ?

srkreddy1238 commented on golang/src/gotvm.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+// GoTVMVersion is gotvm package version information.
+var GoTVMVersion            = "0.1"
```
TVMVersion is already exported below. Hence removing this additional GoTVMVersion.

srkreddy1238 commented on golang/src/value.go:
```diff
@@ -0,0 +1,356 @@
+
+// deleteTVMValue free the native Value object which is allocated in newTVMValue.
+func (tvmval Value) deleteTVMValue() {
+
```
no nothing like that. Just a habit to separate each block.  I will clean the extras here :)

srkreddy1238 commented on golang/src/gotvm.h:
```diff
@@ -0,0 +1,109 @@
+extern void* _native_malloc(int len);
+
+// Version
+extern _gostring_ _TVM_VERSION(void);
```
:+1: 
Most of the places looks like works well with CString and GoString API.
I will enhance to use CFFI across.

srkreddy1238 replied:
Done

tqchen commented:
We can merge that into the cpu docker for now

tqchen commented:
I take another look at the PR. My high level complain is still that we need to minimize gotvm.h and gotvm.cc. So that we can directly call into functions like TVMModGetFunction, and do the free of the created CString on the Go side.

It is always confusing that certain things have to be handled by the caller. I can understand that there might be need for introducing a union type, but still in that case, it is likely that we could simply reinterpret a piece of aligned memory as the union object itself(c.f. https://golang.org/cmd/cgo/ C's union type are represented by byte array with same length) (you can likely do memcpy of data content into it.

srkreddy1238 commented:
Minimizing gotvm.h interface is definitely possible as most of the places wrapper is a direct call.

Reason, I wrapped still is to keep them in one place which helps to add call flow debugging if required later.

Please confirm on this to minimise the wrapper as much as possible.

tqchen commented:
I think it is good to have a clear distinguishment between the TVM C FFI and the necessary wrapper. Minimizing these code will likely makes maintenance easier, if we need some debugging, we can still do so in the tvm runtime

srkreddy1238 commented:
cc @tqchen for review.

tqchen commented on golang/src/array.go:
```diff
@@ -0,0 +1,329 @@
+//#include "gotvm.h"
+import "C"
+
+import (
```
array->ndarray

srkreddy1238 replied:
Done.

tqchen commented:
@srkreddy1238 I have made some comments. Please do remove the deploy.so from commit (codebase should never contain binary blobs) instead, use a python script to generate the necessary shared library

srkreddy1238 commented:
cc @tqchen

tqchen commented:
I tried the golang docker in docker/Dockerfile.ci_cpu but seems that go cannot be found, @srkreddy1238 can you confirm again if the docker script is correct locally?

srkreddy1238 commented:
Sure, will check again

srkreddy1238 commented:
cc @tqchen  https://github.com/dmlc/tvm/pull/2023 should fix it.

srkreddy1238 commented:
cc @tqchen

tqchen commented:
Unfortunately, I didn't have time to look at this recently. We do need an extra reviewer to merge this in. Let us see if there are anyone who is familiar with CFFI that can do the job. @grwlf maybe?

grwlf commented:
Sure!

grwlf commented on golang/src/gotvm.cc:
```diff
@@ -0,0 +1,184 @@
+  char **names_array;
+  int result;
+
+  result = TVMFuncListGlobalNames(&names_size, (char const ***)&names_array);
```
I'd suggest to return immediately if `result!=0`  (ignoring the strange fact that TVMFuncListGlobalNames always return 0))

srkreddy1238 replied:
Done

grwlf commented on golang/src/gotvm.cc:
```diff
@@ -0,0 +1,184 @@
+void _TVMValueNativeSet(void* to_ptr, void* from_ptr, int ind) {
+  TVMValue **from_p = reinterpret_cast<TVMValue**>(from_ptr);
+  TVMValue *to_p = reinterpret_cast<TVMValue*>(to_ptr);
+  memcpy(&(to_p[ind]), from_p, sizeof(TVMValue));
```
I have trouble parsing this `memcpy`. Do we copy TVMValues, as suggested by `sizeof(TVMValue)`? If so, why the `from_p` is declared as pointer to pointer to TVMValue?

srkreddy1238 replied:
from_p represents an array of TVMValue and the destination index is ind, hence a double pointer.

grwlf replied:
In C, double pointer may represent pointer to a single pointer (8 bytes on x86-64) or pointer to an array of pointers (array of 8 bytes uintptrs). AFAIK, none of those readings represents our case where we copy TVMValue to the specific place in array of TVMValues. Numerically this code is correct, but I think that usage of double-pointer is misleading here. Could you please re-check?

srkreddy1238 replied:
Ok. I simplified with single pointers instead.

grwlf commented on golang/src/gotvm.cc:
```diff
@@ -0,0 +1,184 @@
+ */
+void _TVMValueNativeGet(void* to_ptr, void* from_ptr, int ind) {
+  TVMValue *from_p = reinterpret_cast<TVMValue*>(from_ptr);
+  TVMValue **to_p = reinterpret_cast<TVMValue**>(to_ptr);
```
Same as above: should it be a plain pointer to TVMValue, like
`TVMValue *to_p = reinterpret_cast<TVMValue*>(to_ptr);`
?

srkreddy1238 replied:
same as above.

srkreddy1238 commented on golang/src/gotvm.h:
```diff
@@ -0,0 +1,48 @@
+typedef void* native_voidp;
+
+// golang TVMType
+typedef struct { uint8_t code; uint8_t bits; uint16_t lanes; } _tvmtype_;
```
Thanks. This was an old change not needed anymore. Removing.

srkreddy1238 commented on golang/src/gotvm.cc:
```diff
@@ -0,0 +1,184 @@
+  _gostring_ ret;
+  ret.p = reinterpret_cast<char*>(malloc(l));
+  memcpy(ret.p, p, l);
+  ret.n = l;
```
Here data is list of function names we have globally in TVM. Generally never go out of INTMAX.

srkreddy1238 commented:
@grwlf  thanks for review. Rebased the code to later master. 

Welcome to review further.

grwlf commented on golang/Makefile:
```diff
@@ -0,0 +1,63 @@
+	@export GOPATH=$(GOPATH); \
+	export CGO_CPPFLAGS=$(CGO_CPPFLAGS); \
+	export CGO_CFLAGS=$(CGO_CFLAGS); \
+	export CGO_LDFLAGS=$(CGO_LDFLAGS); \
```
Should you set `CGO_CXXFLAGS` here as well?

srkreddy1238 replied:
Done

grwlf commented on golang/README.md:
```diff
@@ -0,0 +1,96 @@
+
+### Requirements
+
+- go compiler (https://golang.org/) version 0.10 or above.
```
Do we have pre-built images with go support? If not, I think it would be nice to mention that users may include `docker/install/ubuntu_install_golang.sh` rule in their Dockerfiles. I tested that way, it worked, but I had to `export PATH="/usr/lib/go-1.10/bin:$PATH"`. Could you please  mention that too?

srkreddy1238 replied:
Demo CPU docker should support golang if it is build from Dockerfile.ci_cpu. cc @tqchen can confirm on this.

Anyway I added a section ```Docker``` to golang README with above info.

grwlf commented:
@srkreddy1238 I was able to run tests, it worked. Please see my comments regarding Makefile and README.

srkreddy1238 commented:
Thanks @grwlf. Addressed the review comments.

tqchen commented:
OK, if there is no further comments, I will merge this PR in 24 hours. Thanks @grwlf for the reviews.

tqchen commented:
Thanks, @grwlf for the detailed reviews, Thanks,  @srkreddy1238 for patiently improving the code during the review process. This is now merged. Please send in followup PRs to enable the test in CI
